### ISSUE TITLE
Outlook Autorun setSignatureAsync moves focus to mailbox

### ISSUE THREAD

**Issue by vdboots** [OfficeDev/office-js#2163]
## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: Office on the web
* Host [Excel, Word, PowerPoint, etc.]: Outlook
* Office version number: Web
* Operating System: ______
* Browser (if using Office on the web): Edge(Chromium)
 
## Expected behavior
Inserts the signature while user can continue typing mailadres

## Current behavior
In the browser the cursor moves from the "To" field to the mailbody.

## Steps to reproduce
1.  launch script where a email is added. 
2. open new email and start typing a name.

**Comment by nancy-wang** [comment]
thank you for sharing this feedback with our team @vdboots! Connecting you with @exextoc who may be able to help here.

**Comment by exextoc** [comment]
we tried the same scenario which is mentioned above but i am able to see that cursor at To field when setSignatureAsync API is called ..Can you please shared the code snippet for us to try to repro this issue??

**Comment by moodsey211** [comment]
I'm not sure if we have the same scenario but I was able to reproduce this issue by calling setSignatureAsync twice.

```
Office.context.mailbox.item.body.setSignatureAsync(
          html,
          {
            coercionType: Office.CoercionType.Html,
          },
          (asyncResult) => {
            console.log('first call');
          }
      );

...

Office.context.mailbox.item.body.setSignatureAsync(
          html,
          {
            coercionType: Office.CoercionType.Html,
          },
          (asyncResult) => {
            console.log('second call');
          }
      );
```

**Comment by vdboots** [comment]
yes we use it twice. First to set signature from cache. and then a second time(if changed).  from an external source.

**Comment by exextoc** [comment]
Thanks for sharing this detail.

Can you please confirm what this step mean? Are you using [LaunchEvents](https://docs.microsoft.com/en-us/office/dev/add-ins/outlook/autolaunch) which inserts signature "twice" in the email as soon as "New Email" is started?

"1. launch script where a email is added."

**Comment by vdboots** [comment]
1. launch script where a email is added.
2. Get updated information from external source
3. Launch script to update signature.

**Comment by exextoc** [comment]
@vdboots Can you please share a test code to repro this scenario?

**Comment by exextoc** [comment]
This is to inform that we could repro this issue and is currently in our backlog. Unfortunately we do not have any timelines for the same.

**Comment by vdboots** [comment]
hi is there any update on this issue ?

**Comment by exextoc** [comment]
We have already provided a fix for this issue which is already in production. Can you confirm that you are still facing this problem.
Also, you have opened another GitHub issue: "https://github.com/OfficeDev/office-js/issues/2551". Can you confirm that it is duplicate of this issue.

**Comment by vdboots** [comment]
this issue still occours.

https://github.com/OfficeDev/office-js/issues/2551 is different since this one focus on single setSignatureAsync Moves out  mail message at all.. Te focus moves now out of the mail at all. even when you run the setSignatureAsync. one time. The issue above is still going on when you run the command twice.

**Comment by exextoc** [comment]
Could you help us out with more detailed repro steps and a test add-in if possible?
Please send it across to: outlookwebaddins@M365x50631954.onmicrosoft.com

**Comment by ElizabethSamuel-MSFT** [comment]
@exextoc Any update on this? Thanks.

**Comment by BCRaveD** [comment]
We are experiencing the same issue with our add-in when using Office on the web.

We are using setSignatureAsync(), and it can be event driven (depending on settings).
For example, when a new message is composed, it will update or insert a signature for the user if required.
The signature may update with subsequent changes to the message recipients or subject etc, if set to do so.
Both move focus from the user's current activity to the message body.

**Environment**
Platform - Office on the web
Browser - Edge (Chromium)

I have also tested in Microsoft® Outlook® for Microsoft 365 MSO (Version 2207 Build 16.0.15427.20182) 64-bit, and have not been able to reproduce the issue. Focus is unchanged when the signature is updated by our add-in.

**Expected behaviour**
Signature updates using setSignatureAsync() do not result in any change of focus from the user's current activity.
eg If a user is typing an email address into the To: field, they should not be interrupted part way though by a signature update.

**Current behaviour**

When a user composes a new email, the focus may be moved to the message body when setSignatureAsync() updates/adds the initial signature. The user could be part way through typing an email address into the To: field at the time.

It will then move focus to the message body again each time any subsequent updates are made to the signature. This could occur whilst the user is changing recipients, or the subject etc (depending on settings).

Whilst a user is composing an email, it is possible that our add-in will use setSignatureAsync() several times. There would usually be a natural delay between each change though, as they require the user to change something in the email. Updates to the signature would not specifically be consecutive, but may queue up if the user makes a number of quick changes sequentially.

**Steps to reproduce**
1. Run a script which uses setSignatureAsync() to update/add a signature to the current message regularly (this should not require any user action when composing a message).
2. Compose a new email in OWA, ensure the cursor is not placed in the message body (any header section should work, eg the default, To:).
3. Wait for the script to run setSignatureAsync() and update the signature.
- Focus is moved to the message body after the signature is updated

Note: I am unable to reproduce this 100% of the time. Closing or sending the draft, and then creating another may cause it to occur again.

**Comment by exextoc** [comment]
Thank you for reporting this issue of cursor movement. It is in our backlog. We unfortunately have no timelines to share at this point. 
Internal tracking ID- 131538

**Comment by ghelyar** [comment]
Just to provide a reliable repro:

1. Create an add-in with the `OnMessageRecipientsChanged` launch event
2. Create an event handler that waits a short delay (e.g. 1 second) - this is just to simulate something like a network request
3. Call `setSignatureAsync` in the event handler with some data that is easy to see, like the current time or the current set of recipients.
4. In OWA, type a recipient in the "to" box, but hold off on finishing it (don't autocomplete it, or press enter, or press semicolon etc). For example, type `user@example.com` and nothing else.
5. Press semicolon and then hold down another letter on the keyboard, e.g. type `;xxxxxxxxxxxx`. The letter does not matter, it's just to simulate typing a second recipient.

If the signature updates while you're typing the second recipient (or in this case holding down the x button) you will see the caret move into the message, and it will have the first half of the recipient's address in the to box and the second half in the message box.

![hc99Gx6DEP](https://user-images.githubusercontent.com/3225358/189316636-2465869c-e2b6-4221-86d9-f4de0c603657.gif)

manifest.xml:
```xml
...
<LaunchEvents>
  <LaunchEvent Type="OnMessageRecipientsChanged" FunctionName="onMessageRecipientsChangedHandler" />
</LaunchEvents>
...
```

launchevent.js:
```javascript
function onMessageRecipientsChangedHandler(event) {
  setTimeout(() => {
    Office.context.mailbox.item.to.getAsync((getToResult) => {
      Office.context.mailbox.item.body.setSignatureAsync(
        getToResult.value.map(recipient => recipient.emailAddress).join(", "),
        { coercionType: Office.CoercionType.Html },
        (setSignatureResult) => {
          event.completed();
        }
      );
    });
  }, 1000);
}

Office.actions.associate("onMessageRecipientsChangedHandler", onMessageRecipientsChangedHandler);
```

If you wait for the signature to update before typing the second recipient, the bug does not occur (but users won't do this).

**Comment by elrapha** [comment]
I am having the same problem here.
@ghelyar Thanks for the visual repo

**Comment by barclayadam** [comment]
We are seeing customer reports of a similar issue with the focus being stolen when the signature is being inserted by our add-in. This has been reported in at least Firefox, but in their case **not** on Chromium-based browsers

The report says focus can be stolen from the To, CC, Subject and Body fields
